test/gpu/octree/test_approx_nearest.cpp

```diff
@@ -56,94 +56,108 @@
+    cloud_device.upload(cloud->points);
 
-    //gpu build 
-    pcl::gpu::Octree octree_device;                
-    octree_device.setCloud(cloud_device);	    
+    //gpu build
+    pcl::gpu::Octree octree_device;
+    octree_device.setCloud(cloud_device);
     octree_device.build();
-    
+
     //build host octree
+    float host_octree_resolution = 0.05;
     pcl::octree::OctreePointCloudSearch<pcl::PointXYZ> octree_host(host_octree_resolution);
-    octree_host.setInputCloud (cloud_host);    
+    octree_host.setInputCloud (cloud);
     octree_host.addPointsFromInputCloud();
-           
+
     //upload queries
     pcl::gpu::Octree::Queries queries_device;
-    queries_device.upload(data.queries);
-    
-        
+    queries_device.upload(queries);
+    //pcl::gpu::Octree::ResultSqrDists distances_device(queries.size());
+
     //prepare output buffers on device
-    pcl::gpu::NeighborIndices result_device(data.tests_num, 1);
-    std::vector<int> result_host_pcl(data.tests_num);
-    std::vector<int> result_host_gpu(data.tests_num);
-    std::vector<float> dists_pcl(data.tests_num);
-    std::vector<float> dists_gpu(data.tests_num);
-    
+    pcl::gpu::NeighborIndices result_device(queries.size(), 1);
+    std::vector<int> result_host_pcl(queries.size());
+    std::vector<int> result_host_gpu(queries.size());
+    std::vector<float> dists_pcl(queries.size());
+    std::vector<float> dists_gpu(queries.size());
+
     //search GPU shared
     octree_device.approxNearestSearch(queries_device, result_device);
-
     std::vector<int> downloaded;
+    std::vector<float>distances;
     result_device.data.download(downloaded);
-                
-    for(std::size_t i = 0; i < data.tests_num; ++i)
+    //distances_device.download(dists_gpu_direct);
+
+    for(size_t i = 0; i < queries.size(); ++i)
     {
-        octree_host.approxNearestSearch(data.queries[i], result_host_pcl[i], dists_pcl[i]);
-        octree_device.approxNearestSearchHost(data.queries[i], result_host_gpu[i], dists_gpu[i]);
+        octree_host.approxNearestSearch(queries[i], result_host_pcl[i], dists_pcl[i]);
+        octree_device.approxNearestSearchHost(queries[i], result_host_gpu[i], dists_gpu[i]);
     }
 
-    ASSERT_EQ ( ( downloaded == result_host_gpu ), true );
+    ASSERT_EQ ((downloaded == result_host_gpu), true);
 
-    int count_gpu_better = 0;
-    int count_pcl_better = 0;
-    float diff_pcl_better = 0;
-    for(std::size_t i = 0; i < data.tests_num; ++i)
+    //find inconsistencies with gpu and cpu cuda impementation
+    //int count_gpu_better = 0;
+    //int count_pcl_better = 0;
+    //int count_different = 0;
+    for(size_t i = 0; i < queries.size(); ++i)
     {
-        float diff = dists_pcl[i] - dists_gpu[i];
-        bool gpu_better = diff > 0;
-
-        ++(gpu_better ? count_gpu_better : count_pcl_better);
-
-        if (!gpu_better)
-            diff_pcl_better +=std::abs(diff);
+        ASSERT_EQ ((dists_pcl[i] == dists_gpu[i]), true);
     }
```

Comment:
```diff
diff --git a/test/gpu/octree/test_approx_nearest.cpp b/test/gpu/octree/test_approx_nearest.cpp
index c4b0caea9..2d0c9f879 100644
--- a/test/gpu/octree/test_approx_nearest.cpp
+++ b/test/gpu/octree/test_approx_nearest.cpp
@@ -35,137 +35,141 @@
  */
 
 #if defined _MSC_VER
-    #pragma warning (disable : 4996 4530)
+#pragma warning(disable : 4996 4530)
 #endif
 
 #include <gtest/gtest.h>
 
-#include <iostream>
-#include <fstream>
 #include <algorithm>
+#include <fstream>
+#include <iostream>
 
 #if defined _MSC_VER
-    #pragma warning (disable: 4521)
+#pragma warning(disable : 4521)
 #endif
-#include <pcl/point_cloud.h>
 #include <pcl/octree/octree_search.h>
+#include <pcl/point_cloud.h>
 #if defined _MSC_VER
-    #pragma warning (default: 4521)
+#pragma warning(default : 4521)
 #endif
 
-#include <pcl/gpu/octree/octree.hpp>
 #include <pcl/gpu/containers/device_array.h>
+#include <pcl/gpu/octree/octree.hpp>
 
 using namespace pcl::gpu;
 
-//TEST(PCL_OctreeGPU, DISABLED_approxNearesSearch)
+// TEST(PCL_OctreeGPU, DISABLED_approxNearesSearch)
 TEST(PCL_OctreeGPU, approxNearesSearch)
 {
-    //generate custom pointcloud
-    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
-
-    // Fill in cloud data
-    const pcl::index_t point_size = 10000;
-    cloud->width    = point_size;
-    cloud->height   = 1;
-    cloud->is_dense = false;
-    cloud->points.resize (cloud->width * cloud->height);
-
-    /*
-    //the test points create an octree with bounds (-1, -1, -1) and (1, 1, 1).
-    ------------------------------------
-    |                |                 |
-    |                |                 |
-    |                |                 |
-    |                |                 |
-    |                |                 |
-    |----------------------------------|
-    | x     | q      |                 |
-    |       |        |                 |
-    |-------|--------|                 |
-    |       | y      |                 |
-    |       |        |                 |
-    ------------------------------------
-    the final two point are positioned such that point 'x' is father from query point 'q' than 'y',
-    but the voxel containing 'x' is closer to  'q' than the voxel containing 'y'
-    */
-    const float x_cords[point_size] = {-1, -1, -1, 1, -1, 1, 1, 1, -0.9, -0.4};
-    const float y_cords[point_size] = {-1, -1, 1, -1, 1, -1, 1, 1, -0.2, -0.6};
-    const float z_cords[point_size] = {-1, 1, -1, -1, 1, 1, -1, 1, -0.75, -0.75};
-
-    for (std::size_t i = 0; i < cloud->size (); ++i)
-    {
-        (*cloud)[i].x = x_cords[i%10];
-        (*cloud)[i].y = y_cords[i%10];
-        (*cloud)[i].z = z_cords[i%10];
-    }
-
-    std::vector<pcl::PointXYZ> queries;
-    //While the GPU implementation has a fixed depth of 10 levels, octree depth in the CPU implementation can vary based on
-    //the leaf size set by the user, which can affect the results. Therefore results would only tally if depths match.
-    queries.push_back(pcl::PointXYZ(-0.4, -0.2, -0.75));     //should be different across CPU and GPU if different traversal methods are used
-    queries.push_back(pcl::PointXYZ(-0.6, -0.2, -0.75));     //should be same across CPU and GPU
-    queries.push_back(pcl::PointXYZ(1.1, 1.1, 1.1));         //out of range query
-
-    //prepare device cloud
-    pcl::gpu::Octree::PointCloud cloud_device;
-    cloud_device.upload(cloud->points);
-
-    //gpu build
-    pcl::gpu::Octree octree_device;
-    octree_device.setCloud(cloud_device);
-    octree_device.build();
-
-    //build host octree
-    float host_octree_resolution = 0.05;
-    pcl::octree::OctreePointCloudSearch<pcl::PointXYZ> octree_host(host_octree_resolution);
-    octree_host.setInputCloud (cloud);
-    octree_host.addPointsFromInputCloud();
-
-    //upload queries
-    pcl::gpu::Octree::Queries queries_device;
-    queries_device.upload(queries);
-    //pcl::gpu::Octree::ResultSqrDists distances_device(queries.size());
-
-    //prepare output buffers on device
-    pcl::gpu::NeighborIndices result_device(queries.size(), 1);
-    std::vector<int> result_host_pcl(queries.size());
-    std::vector<int> result_host_gpu(queries.size());
-    std::vector<float> dists_pcl(queries.size());
-    std::vector<float> dists_gpu(queries.size());
-
-    //search GPU shared
-    octree_device.approxNearestSearch(queries_device, result_device);
-    std::vector<int> downloaded;
-    std::vector<float>distances;
-    result_device.data.download(downloaded);
-    //distances_device.download(dists_gpu_direct);
-
-    for(size_t i = 0; i < queries.size(); ++i)
-    {
-        octree_host.approxNearestSearch(queries[i], result_host_pcl[i], dists_pcl[i]);
-        octree_device.approxNearestSearchHost(queries[i], result_host_gpu[i], dists_gpu[i]);
-    }
-
-    ASSERT_EQ ((downloaded == result_host_gpu), true);
-
-    //find inconsistencies with gpu and cpu cuda impementation
-    //int count_gpu_better = 0;
-    //int count_pcl_better = 0;
-    //int count_different = 0;
-    for(size_t i = 0; i < queries.size(); ++i)
-    {
-        ASSERT_EQ ((dists_pcl[i] == dists_gpu[i]), true);
-    }
 
+  /*
+  the test points create an octree with bounds (-1, -1, -1) and (1, 1, 1).
+  point q, represents a query point
+  ------------------------------------
+  |                |                 |
+  |                |                 |
+  |                |                 |
+  |                |                 |
+  |                |                 |
+  |----------------------------------|
+  | x     | q      |                 |
+  |       |        |                 |
+  |-------|--------|                 |
+  |       | y      |                 |
+  |       |        |                 |
+  ------------------------------------
+  the final two point are positioned such that point 'x' is father from query point 'q'
+  than 'y', but the voxel containing 'x' is closer to  'q' than the voxel containing 'y'
+  */
+
+  // std::array requires the explicity call to pcl::PointXYZ's ctor
+  const std::array<pcl::PointXYZ, 10> coords{
+      pcl::PointXYZ{-1.f, -1.f, -1.f},
+      pcl::PointXYZ{-1.f, -1.f, 1.f},
+      pcl::PointXYZ{-1.f, 1.f, -1.f},
+      pcl::PointXYZ{-1.f, 1.f, 1.f},
+      pcl::PointXYZ{1.f, -1.f, -1.f},
+      pcl::PointXYZ{1.f, -1.f, 1.f},
+      pcl::PointXYZ{1.f, 1.f, -1.f},
+      pcl::PointXYZ{1.f, 1.f, 1.f},
+      pcl::PointXYZ{-0.9f, -0.2f, -0.75f},
+      pcl::PointXYZ{-0.4f, -0.6f, -0.75f},
+  };
+
+  // While the GPU implementation has a fixed depth of 10 levels, octree depth in the
+  // CPU implementation can vary based on the leaf size set by the user, which can affect
+  // the results. Therefore results would only tally if depths match. generate custom
+  // pointcloud
+  constexpr pcl::index_t point_size = 1000 * coords.size();
+  auto cloud = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>(point_size, 1);
+
+  // copy chunks of 10 points at the same time
+  for (auto it = cloud->begin(); it != cloud->cend(); it += coords.size())
+    std::copy(coords.cbegin(), coords.cend(), it);
+
+  // std::vector doesn't
+  const std::vector<pcl::PointXYZ> queries = {
+      {-0.4, -0.2, -0.75}, // should be different across CPU and GPU if different
+                           // traversal methods are used
+      {-0.6, -0.2, -0.75}, // should be same across CPU and GPU
+      {1.1, 1.1, 1.1},     // out of range query
+  };
+
+  // prepare device cloud
+  pcl::gpu::Octree::PointCloud cloud_device;
+  cloud_device.upload(cloud->points);
+
+  // gpu build
+  pcl::gpu::Octree octree_device;
+  octree_device.setCloud(cloud_device);
+  octree_device.build();
+
+  // build host octree
+  constexpr float host_octree_resolution = 0.05;
+  pcl::octree::OctreePointCloudSearch<pcl::PointXYZ> octree_host(
+      host_octree_resolution);
+  octree_host.setInputCloud(cloud);
+  octree_host.addPointsFromInputCloud();
+
+  // upload queries
+  pcl::gpu::Octree::Queries queries_device;
+  queries_device.upload(queries);
+  // pcl::gpu::Octree::ResultSqrDists distances_device(queries.size());
+
+  // prepare output buffers on device
+  pcl::gpu::NeighborIndices result_device(queries.size(), 1);
+  std::vector<int> result_host_pcl(queries.size());
+  std::vector<int> result_host_gpu(queries.size());
+  std::vector<float> dists_pcl(queries.size());
+  std::vector<float> dists_gpu(queries.size());
+
+  // search GPU shared
+  octree_device.approxNearestSearch(queries_device, result_device);
+  std::vector<int> downloaded;
+  result_device.data.download(downloaded);
+
+  for (size_t i = 0; i < queries.size(); ++i) {
+    octree_host.approxNearestSearch(queries[i], result_host_pcl[i], dists_pcl[i]);
+    octree_device.approxNearestSearchHost(queries[i], result_host_gpu[i], dists_gpu[i]);
+  }
+
+  ASSERT_EQ(downloaded, result_host_gpu);
+
+  // find inconsistencies with gpu and cpu cuda impementation
+  // int count_gpu_better = 0;
+  // int count_pcl_better = 0;
+  // int count_different = 0;
+  for (size_t i = 0; i < queries.size(); ++i) {
+    ASSERT_EQ(dists_pcl[i], dists_gpu[i]);
+    ASSERT_EQ(dists_pcl[i], dists_gpu[i]);
+  }
 }
 
 /* ---[ */
 int
-main (int argc, char** argv)
+main(int argc, char** argv)
 {
-  testing::InitGoogleTest (&argc, argv);
-  return (RUN_ALL_TESTS ());
+  testing::InitGoogleTest(&argc, argv);
+  return (RUN_ALL_TESTS());
 }
 /* ]--- */
-
```

Reply:
Please remove also all the windows macros disabling warnings at the beginning of this file. Those should be left on so that we can handle them later.